### PR TITLE
Update demos to remove warnings

### PR DESCRIPTION
**Pull request by mudit2812**
* Added `interface=autograd` during qnode initialization of tutorials that were generating warnings due to device switching.
* Replaced references to `qml.utils.sparse_hamiltonian` with `qml.Hamiltonian.sparse_matrix`.
* Removed references to `QNodeCollection`.

### CODE REVIEW

File: demonstrations/tutorial_noisy_circuit_optimization.py [hunk 1]
```diff
@@ -94,39 +94,22 @@
...
-    bell_pair()
-    return qml.expval(A2 @ B1)
 
+# circuits for measuring each distinct observable
 @qml.qnode(dev)
-def measure_A2B2():
+def measure_expvals():
     bell_pair()
-    return qml.expval(A2 @ B2)
+    return qml.expval(A1 @ B1), qml.expval(A1 @ B2), qml.expval(A2 @ B1), qml.expval(A2 @ B2)
```
Comment (rmoyard) [inline demonstrations/tutorial_noisy_circuit_optimization.py:108]:
```suggestion
    return [qml.expval(A1 @ B1), qml.expval(A1 @ B2), qml.expval(A2 @ B1), qml.expval(A2 @ B2)]
```

File: demonstrations/tutorial_noisy_circuit_optimization.py [hunk 2]
```diff
@@ -94,39 +94,22 @@
...
 
-circuits = qml.QNodeCollection([measure_A1B1,
-                                measure_A1B2,
-                                measure_A2B1,
-                                measure_A2B2])
 
-# now we measure each circuit and construct the CHSH inequality
-expvals = circuits()
+# now we measure each observable and construct the CHSH inequality
+expvals = list(measure_expvals())
```
Comment (rmoyard) [inline demonstrations/tutorial_noisy_circuit_optimization.py:112]: I think you can add the list above directly

File: demonstrations/tutorial_noisy_circuit_optimization.py [hunk 3]
```diff
@@ -452,17 +438,18 @@ def noisy_cost(x):
...
 
-plt.plot(angles, noisy_expvals,
-         label="Expectation value")  # looks like 0.4 * cos(phi)
-plt.plot(angles, noisy_param_shift,
-         label="Parameter-shift value")  # looks like -0.4 * sin(phi)
+plt.plot(angles, noisy_expvals, label="Expectation value")  # looks like 0.4 * cos(phi)
+plt.plot(angles, noisy_param_shift, label="Parameter-shift value")  # looks like -0.4 * sin(phi)
 plt.ylabel(r"Expectation value $\langle \hat{Z} \rangle$")
 plt.xlabel(r"Angle $\theta_1$")
 plt.legend()
```
Comment (rmoyard) [inline demonstrations/tutorial_noisy_circuit_optimization.py:455]: General question how are the plots compared to master?

**Review by rmoyard — COMMENTED** [review]
Thanks @mudit2812 for this tedious task! Just a couple of questions for the removal of QNode collection

**Comment by rmoyard** [comment]
In `tutorial_noisy_circuit_optimization` we run into trouble: 
```
 TypeError: Expected a Pauli word Observable instance, instead got expval(PauliZ(wires=[0]) @ Hermitian(array([[ 0.70710678,  0.70710678],
       [ 0.70710678, -0.70710678]]), wires=[1])).
```

**Comment by mudit2812** [comment]
`default.qubit`'s batch transform is trying to split the circuit into multiple tapes and failing. I'll just use a list of the results of each qnode.

File: demonstrations/tutorial_noisy_circuit_optimization.py (hunk 3 above)
Comment (mudit2812) [reply demonstrations/tutorial_noisy_circuit_optimization.py:473]: They look basically the same

**Review by rmoyard — APPROVED** [review]
Great thanks :100:

**Review by lillian542 — APPROVED** [review]
Looks good to me!

**Pull request merged**